lightning-data-persister/src/lib.rs

```diff
@@ -0,0 +1,239 @@
+use std::fs;
+use std::io::{Error, ErrorKind, Cursor};
+use std::marker::PhantomData;
+use std::path::Path;
+
+#[cfg(not(target_os = "windows"))]
+use std::os::unix::io::AsRawFd;
+
+/// FilesystemPersister can persist channel data on disk on Linux machines, where
+/// each channel's data is stored in a file named after its funding outpoint.
+///
+/// Warning: this module does the best it can with calls to persist data, but it
+/// can only guarantee that the data is passed to the drive. It is up to the
+/// drive manufacturers to do the actual persistence properly, which they often
+/// don't (especially on consumer-grade hardware). Therefore, it is up to the
+/// user to validate their entire storage stack, to ensure the writes are
+/// persistent.
+/// Corollary: especially when dealing with larger amounts of money, it is best
+/// practice to have multiple channel data backups and not rely only on the
+/// FilesystemPersister.
+pub struct FilesystemPersister<ChanSigner: ChannelKeys + Readable + Writeable> {
+	path_to_channel_data: String,
+	phantom: PhantomData<ChanSigner>, // TODO: is there a way around this?
+}
+
+impl<ChanSigner: ChannelKeys + Readable + Writeable> FilesystemPersister<ChanSigner> {
+	/// Initialize a new FilesystemPersister and set the path to the individual channels'
+	/// files.
+	pub fn new(path_to_channel_data: String) -> Self {
+		return Self {
+			path_to_channel_data,
+			phantom: PhantomData,
+		}
+	}
+
+	fn get_full_filepath(&self, funding_txo: OutPoint) -> String {
+		let dir_path = Path::new(&self.path_to_channel_data);
+		let mut path_buf = dir_path.to_path_buf();
+		path_buf.push(format!("{}_{}", funding_txo.txid.to_hex(), funding_txo.index));
+		path_buf.to_str().unwrap().to_string()
+	}
+
+	// Utility to write a file to disk.
+	// Note: we may eventually want to change this model to a queue-based system, such that
+	// writes are queued to a separate writer thread. This would improve performance
+	// since syncing to disk can sometimes take 100s of milliseconds.
+	fn write_channel_data(&self, funding_txo: OutPoint, monitor: &ChannelMonitor<ChanSigner>) -> std::io::Result<()> {
+		// Do a crazy dance with lots of fsync()s to be overly cautious here...
+		// We never want to end up in a state where we've lost the old data, or end up using the
+		// old data on power loss after we've returned.
+		// The way to atomically write a file on Unix platforms is:
+		// open(tmpname), write(tmpfile), fsync(tmpfile), close(tmpfile), fsync(dir), rename(), fsync(dir)
+		let filename = self.get_full_filepath(funding_txo);
+		let tmp_filename = filename.clone() + ".tmp";
+
+		{
+			let mut f = fs::File::create(&tmp_filename)?;
+			monitor.write_for_disk(&mut f)?;
+			f.sync_all()?;
+		}
+		// We don't need to create a backup if didn't already have the file, but in any other case
+		// try to create the backup and expect failure on fs::copy() if eg there's a perms issue.
+		let need_bk = match fs::metadata(&filename) {
```

Comment:
it seems that we will always be in a consistent state, before or after the rename below, so there doesn't seem to be a need for a backup.

Reply:
I'm unsure about this part too. @TheBlueMatt why is a backup file necessary? (this was copied from RLBRPC)

Reply:
I can't find any evidence that rename is guaranteed to be atomic across power failure, and in fact I believe it was *not* atomic on old versions of some filesystems, though it may be today. ie its totally possible to write the new file, sync it, sync the world, call rename, and crash and end up with no file.

Reply:
Ok, I stand corrected, I'm not actually sure why I thought this initially, but I can't find any references to any bugs which would make this matter, I agree it should be replaced.